torch/csrc/autograd/python_function.cpp

```diff
@@ -1055,3 +1069,67 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
   Py_INCREF((PyObject*)self);
   return std::shared_ptr<PyFunction>(&self->cdata, Decref());
 }
+
+namespace torch { namespace autograd {
+
+using memo_map = std::unordered_map<Node*, variable_list>;
+
+variable_list interpret_node(std::shared_ptr<Node>, input_map&, memo_map&);
+
+std::shared_ptr<Variable> interpret_output(Output output, input_map& inputs, memo_map& memo) {
+    auto vs = interpret_node(output.node, inputs, memo);
+    return vs.at(output.output_nr);
+}
+
+variable_list interpret_node(std::shared_ptr<Node> node, input_map& inputs, memo_map& memo) {
+    variable_list input_vars;
+    auto num_args = node->inputs.size();
+    input_vars.reserve(num_args);
+    for (auto input : node->inputs) {
+        input_vars.emplace_back(interpret_output(input, inputs, memo));
```

Comment:
I don't think a recursive solution is a good idea. It might overflow the stack once the graph becomes very deep (say ResNet-1000)